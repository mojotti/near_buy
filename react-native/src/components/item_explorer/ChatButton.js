import React from 'react';
import { ActivityIndicator, Text, TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import { styles } from '../../static/styles/ItemDetailsStyles';


export default class ChatButton extends React.Component {
  _getIcon = () => {
    if (this.props.isCreatingChat) {
      return <ActivityIndicator size="large" color="#FFFFFF" />;
    }
    return <FontAwesome name="comments" size={50} color="#FFFFFF" />;
  };

  _getText = () => {
    if (this.props.isCreatingChat) {
      return 'Creating chat...';
    }
    return 'Chat with owner';
  };

  render() {
    return (
      <View>
        <TouchableOpacity
          style={styles.chatButton}
          onPress={this.props.onPress}
        >
          <View style={styles.chatButtonContainer}>
            {this._getIcon()}
            <Text style={styles.chatText}>{this._getText()}</Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
}

ChatButton.propTypes = {
  onPress: PropTypes.func.isRequired,
  isCreatingChat: PropTypes.bool.isRequired,
};
