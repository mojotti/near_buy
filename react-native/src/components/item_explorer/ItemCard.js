import React from 'react';
import { Image, Text, TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { styles } from '../../static/styles/ItemCardStyles';
import { baseStyles } from '../../static/styles/BaseStyles';
import { localhost } from '../../static/constants';

export class _ItemCard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      distanceInKm: 'Getting distance...',
    };
  }

  componentDidMount() {
    this._getDistanceInKm(
      this.props.latitude,
      this.props.longitude,
      this.props.item.latitude,
      this.props.item.longitude,
    );
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.item.latitude !== this.props.item.latitude ||
      prevProps.item.longitude !== this.props.item.longitude ||
      prevProps.latitude !== this.props.latitude ||
      prevProps.longitude !== this.props.longitude
    ) {
      this._getDistanceInKm(
        this.props.latitude,
        this.props.longitude,
        this.props.item.latitude,
        this.props.item.longitude,
      );
    }
  }

  _degreesToRadians = degrees => degrees * Math.PI / 180;

  _getDistanceInKm = (lat1, lon1, lat2, lon2) => {
    const R = 6371; // Radius of the earth in km
    const dLat = this._degreesToRadians(lat2 - lat1);
    const dLon = this._degreesToRadians(lon2 - lon1);
    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this._degreesToRadians(lat1)) *
        Math.cos(this._degreesToRadians(lat2)) *
        Math.sin(dLon / 2) *
        Math.sin(dLon / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const d = R * c; // Distance in km

    this.setState(() => ({ distanceInKm: `${d.toFixed(1)} km` }));
  };

  _getImageSource = () =>Â {
    return `http://${localhost}:5000/api/v1.0/${this.props.item.id}/image0.jpg`;
  };

  _navigateToItem = () => {
    const { navigate } = this.props.navigation;
    navigate(
      'ItemDetails',
      { item: this.props.item, distance: this.state.distanceInKm },
    );
  };

  render() {
    return (
      <View style={styles.container}>
        <TouchableOpacity onPress={this._navigateToItem}>
          <Text style={baseStyles.headerText}>{this.props.item.title}</Text>
          <Image source={{ uri: this._getImageSource() }} style={styles.image} />
          <Text style={baseStyles.headerText}>
            {`Distance: ${this.state.distanceInKm}`}
          </Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const mapStateToProps = state => {
  const { latitude, longitude } = state.locationReducer;
  return { latitude, longitude };
};

const ItemCard = connect(mapStateToProps, null)(_ItemCard);
export default ItemCard;

_ItemCard.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string,
    id: PropTypes.number,
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
  }).isRequired,
  latitude: PropTypes.number.isRequired,
  longitude: PropTypes.number.isRequired,
};
