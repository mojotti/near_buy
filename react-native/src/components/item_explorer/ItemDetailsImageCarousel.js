import React from 'react';
import { Dimensions } from 'react-native';
import PropTypes from 'prop-types';
import Carousel from 'react-native-looped-carousel';
import ImagePlaceholder from '../image_placeholder/ImagePlaceholder';
import { styles } from '../../static/styles/ItemDetailsImageCarouselStyles';

const { width } = Dimensions.get('window');


export default class ItemDetailsImageCarousel extends React.Component {
  render() {
    return (
      <Carousel
        style={styles.size}
        autoplay={false}
        pageInfo
        currentPage={0}
        pageInfoBackgroundColor={'#292929'}
        pageInfoTextStyle={{ color: '#FFFFFF' }}
      >
        {this.props.images.map((uri, i) => (
          <ImagePlaceholder
            url={uri}
            styles={styles.size}
            placeholderSize={width}
            key={i}
          />
        ))}
      </Carousel>
    );
  }
}

ItemDetailsImageCarousel.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
};
