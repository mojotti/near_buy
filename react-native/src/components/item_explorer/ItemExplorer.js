'use strict';

import React from 'react';
import { ListView, Text, TouchableOpacity, View } from 'react-native';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/MaterialIcons';

import { styles } from '../../static/styles/ItemsStyles';

export class ItemExplorer extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const { navigate } = navigation;
    return {
      title: 'Item Explorer',
      headerLeft: (
        <View style={{ paddingHorizontal: 10 }}>
          <TouchableOpacity onPress={() => navigation.navigate('DrawerOpen')}>
            <Icon name="menu" size={30} color="blue" />
          </TouchableOpacity>
        </View>
      ),
      headerRight: (
        <Text onPress={() => navigate('NewItem')} style={styles.newItem}>
          New item
        </Text>
      ),
    };
  };

  render() {
    return (
      <View>
        <Text>Here will be items</Text>
      </View>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {};
};

export default connect(mapStateToProps, null)(ItemExplorer);
