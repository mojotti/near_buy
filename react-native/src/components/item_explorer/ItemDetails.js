import React from 'react';
import { Text, ScrollView, View } from 'react-native';
import { connect } from 'react-redux';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import ItemDetailsImageCarousel from './ItemDetailsImageCarousel';
import { getNumOfPictures } from '../../networking/ApiCalls';
import { localhost } from '../../static/constants';
import { baseStyles } from '../../static/styles/BaseStyles';
import { styles } from '../../static/styles/ItemDetailsStyles';
import ChatButton from './ChatButton';
import ItemSeparator from '../user_items/ItemSeparator';
import { createChatAction } from '../../redux/actions/ChatActions';

window.navigator.userAgent = 'ReactNative';

const io = require('socket.io-client/dist/socket.io');
const connectionConfig = {
  jsonp: false,
  reconnection: true,
  reconnectionDelay: 100,
  reconnectionAttempts: 100000,
  transports: ['websocket'], // you need to explicitly tell it to use websockets
};

const path = `http://${localhost}:5000`;

export class _ItemDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      imageUrls: ['foo.bar'], // provide some uri while loading actual img uris
    };

    this.socket = null;
    if (!this.socket || !this.socket.connected) {
      this.socket = io(path, connectionConfig);
      this.socket.on('connect', () => {
        console.log('connected!');
      });
      this.socket.emit('lol', 'Hello world!');
    }
  }

  static navigationOptions = ({ navigation }) =>
    ({ title: navigation.state.params.item.title });

  componentWillMount() {
    this.getImageUrls();
  }

  getImageUrls = () => {
    getNumOfPictures(this.props.item.id, this.props.token)
      .then((numOfPics) => {
        this.setImageUrls(numOfPics);
      });
  };

  setImageUrls = (numOfPics) => {
    const urls = [];
    for (let i = 0; i < numOfPics; i++) {
      const imagePath =
        `http://${localhost}:5000/api/v1.0/${this.props.item.id}/image${i}.jpg`;
      urls.push(imagePath);
    }
    this.setState(() => ({ imageUrls: urls }));
  };

  _openChat = () => {
    this.props.dispatch(createChatAction(
      this.props.item.seller_id,
      this.props.item.id,
      this.props.token,
    ));
  };

  render() {
    return (
      <ScrollView style={styles.container}>
        <ItemDetailsImageCarousel images={this.state.imageUrls} />
        <Text style={baseStyles.headerText}>Distance</Text>
        <View style={styles.locationIconContainer}>
          <FontAwesome name="map-marker" size={20} color={'#9f9f9f'} style={styles.locationIcon} />
          <Text style={styles.plainText}>{this.props.distance}</Text>
        </View>
        <ItemSeparator widthPercentage={0.86} />
        <Text style={baseStyles.headerText}>Description</Text>
        <Text style={styles.plainText}>{this.props.item.description}</Text>
        <ItemSeparator widthPercentage={0.86} />
        <ChatButton onPress={this._openChat} isCreatingChat={this.props.isCreatingChat} />
      </ScrollView>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const { item } = ownProps.navigation.state.params;
  const { distance } = ownProps.navigation.state.params;
  const { token } = state.authorizationReducer;
  const { isLoading } = state.chatCreationReducer;

  return {
    item,
    token,
    distance,
    isCreatingChat: isLoading,
  };
};

const ItemDetails = connect(mapStateToProps, null)(_ItemDetails);
export default ItemDetails;
