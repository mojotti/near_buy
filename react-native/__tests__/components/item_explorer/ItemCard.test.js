import React from 'react';
import { shallow } from 'enzyme';
import { _ItemCard } from '../../../src/components/item_explorer/ItemCard';

const ITEM = {
  item: {
    title: 'foo',
    id: 0,
  },
  latitude: 0,
  longitude: 0,
};
const LOCATION = {
  lat: 15,
  lon: 100,
};

describe('<_ItemCard />', () => {
  test('transfers degrees to radians', () => {
    const oneDegreeToRadian = 0.0174532925; // accuracy of ten decimals

    const wrapper = shallow(
      <_ItemCard
        item={ITEM}
        latitude={LOCATION.lat}
        longitude={LOCATION.lon}
      />
    );
    const result = wrapper.instance()._degreesToRadians(1);

    expect(eval(result.toFixed(10))).toEqual(oneDegreeToRadian);
  });

  test('gets distance in km between two coordinates', () => {
    const ouluHelsinkiDistance = '534.6 km';
    const oulu = { lat: 65, lon: 25 }; // not really but almost
    const helsinki = { lat: 60.192059, lon: 24.945831 };

    const wrapper = shallow(
      <_ItemCard
        item={ITEM}
        latitude={LOCATION.lat}
        longitude={LOCATION.lon}
      />
    );
    wrapper
      .instance()
      ._getDistanceInKm(oulu.lat, oulu.lon, helsinki.lat, helsinki.lon);

    expect(wrapper.state('distanceInKm')).toEqual(ouluHelsinkiDistance);
  });

  test('navigates to item on press', () => {
    const naviSpy = jest.fn();
    const navi = {
      navigate: naviSpy
    };
    const wrapper = shallow(
      <_ItemCard
        item={ITEM}
        latitude={LOCATION.lat}
        longitude={LOCATION.lon}
        navigation={navi}
      />
    );
    const touchableArea = wrapper.find('TouchableOpacity');
    touchableArea.simulate('Press');

    expect(naviSpy.mock.calls).toMatchSnapshot();
    expect(naviSpy.mock.calls.length).toEqual(1);
  });

  test('component did update updates distance when prop differs', () => {
    const wrapper = shallow(
      <_ItemCard
        item={ITEM}
        latitude={LOCATION.lat}
        longitude={LOCATION.lon}
      />
    );

    const getDistanceSpy = jest.spyOn(wrapper.instance(), '_getDistanceInKm')
    const prevProps = {
      item: {
        latitude: 64,
        longitude: 25,
      }
    };
    expect(getDistanceSpy.mock.calls.length).toEqual(0);
    wrapper.instance().componentDidUpdate(prevProps);

    expect(getDistanceSpy.mock.calls.length).toEqual(1);
  });
});
