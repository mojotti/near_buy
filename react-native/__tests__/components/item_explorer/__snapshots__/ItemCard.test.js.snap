// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<_ItemCard /> navigates to item on press 1`] = `
Array [
  Array [
    "ItemDetails",
    Object {
      "distance": "11081.2 km",
      "item": Object {
        "item": Object {
          "id": 0,
          "title": "foo",
        },
        "latitude": 0,
        "longitude": 0,
      },
    },
  ],
]
`;
